consensus/src/round_manager.rs

```diff
@@ -308,46 +310,40 @@ impl<T: Payload> RoundManager<T> {
     /// and fetch all the blocks from the committed state to the HQC
     /// 2. forwarding the proposals to the ProposerElection queue,
     /// which is going to eventually trigger one winning proposal per round
-    async fn pre_process_proposal(&mut self, proposal_msg: ProposalMsg<T>) -> Option<Block<T>> {
+    async fn pre_process_proposal(&mut self, proposal_msg: ProposalMsg<T>) -> Result<Block<T>> {
         trace_event!("round_manager::pre_process_proposal", {"block", proposal_msg.proposal().id()});
         // RoundState is going to be updated with all the proposal certificates later,
         // but it's known that the round_state's round is not going to decrease so we can already
         // filter out the proposals from old rounds.
         let current_round = self.round_state.current_round();
-        if proposal_msg.round() < current_round {
-            return None;
-        }
-        if self
-            .proposer_election
-            .is_valid_proposer(proposal_msg.proposer(), proposal_msg.round())
-            .is_none()
-        {
-            warn!(
-                "Proposer {} for block {} is not a valid proposer for this round",
-                proposal_msg.proposer(),
-                proposal_msg.proposal()
-            );
-            return None;
-        }
-        if let Err(e) = self
-            .sync_up(proposal_msg.sync_info(), proposal_msg.proposer(), true)
+        ensure!(
+            proposal_msg.round() >= current_round,
+            "[RoundManager] Proposal round {} is less than current round {}",
+            proposal_msg.round(),
+            current_round,
+        );
+        ensure!(
+            self.proposer_election
+                .is_valid_proposer(proposal_msg.proposer(), proposal_msg.round())
+                .is_some(),
+            "[RoundManager] Proposer {} for block {} is not a valid proposer for this round",
+            proposal_msg.proposer(),
+            proposal_msg.proposal()
+        );
+        self.sync_up(proposal_msg.sync_info(), proposal_msg.proposer(), true)
             .await
-        {
-            warn!(
-                "Dependencies of proposal {} could not be added to the block store: {}",
-                proposal_msg, e
-            );
-            return None;
-        }
+            .context("[RoundManager] Process proposal")?;
 
         // round_state may catch up with the SyncInfo, check again
         let current_round = self.round_state.current_round();
-        if proposal_msg.round() != current_round {
-            return None;
-        }
+        ensure!(
+            proposal_msg.round() == current_round,
+            "[RoundManager] Proposal round doesn't match current round after sync"
+        );
 
         self.proposer_election
             .process_proposal(proposal_msg.take_proposal())
+            .ok_or_else(|| format_err!("[RoundManager] Not primary proposal"))
```

Comment:
Invalid author for the current round and round state?

Reply:
it's valid unless we remove secondary proposer (which I'd like to do :) )